File: fleet-manager/app/services/server_service.py
# /fleet-manager/app/services/server_service.py
import uuid
import secrets
import os
from fastapi import HTTPException, status
from redis import Redis
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from app.manager import ServerManager
    
from app.repositories.server_repo import ServerRepository
from app.schemas.schemas import ValheimConfigValidator
from app.clients.identity_client import IdentityServiceClient

class ServerService:
    def __init__(
        self, 
        server_repo: ServerRepository,
        manager: "ServerManager", 
        redis: Redis,
        identity_client: IdentityServiceClient
    ):
        self.server_repo = server_repo
        self.manager = manager
        self.redis = redis
        self.identity_client = identity_client

    def list_servers(self, user_id: str):
        """Fetches servers from the database so the UUID is preserved and filtered by owner."""
        
        try:
            user_id_int = int(user_id)
        except ValueError:
            raise HTTPException(status_code=400, detail="Invalid user ID format.")
            
        db_servers = self.server_repo.get_by_owner(user_id_int) 
        
        server_list = []
        for server_record in db_servers:
            is_online = False
            if server_record.active_pod_name:
                container = self.manager.get_container(server_record.active_pod_name)
                is_online = container is not None and container.status == "running"
            
            custom_name = None
            if server_record.config:
                custom_name = server_record.config.get("VALHEIM_SERVER_NAME")
                
            display_name = custom_name if custom_name else f"{server_record.game_id} Server".title()
            
            server_list.append({
                "id": server_record.id,
                "name": display_name,
                "status": "online" if is_online else "offline",
                "cpu": 0, 
                "ram": 0, 
                "players": 0 
            })
            
        return server_list

    def get_server_details(self, server_id: str):
        """Combines K8s container info with Redis stats using the DB as the source of truth."""
        server_record = self.server_repo.get(server_id)
        if not server_record:
            raise HTTPException(status_code=404, detail="Server not found in database")

        container = None
        if server_record.active_pod_name:
            container = self.manager.get_container(server_record.active_pod_name)

        stats_key = f"server_stats:{server_id}"
        stats = self.redis.hgetall(stats_key)

        cpu = float(stats.get(b"cpu", stats.get("cpu", 0)))
        ram = float(stats.get(b"ram", stats.get("ram", 0)))
        players = int(stats.get(b"players", stats.get("players", 0)))

        is_online = container is not None and getattr(container, "status", "") == "running"

        custom_name = None
        if server_record.config:
            custom_name = server_record.config.get("VALHEIM_SERVER_NAME")
            
        display_name = custom_name if custom_name else f"{server_record.game_id} Server".title()

        return {
            "id": server_record.id,
            "name": display_name,
            "status": "online" if is_online else "offline",
            "cpu": cpu if is_online else 0,
            "ram": ram if is_online else 0,
            "players": players if is_online else 0
        }

    async def toggle_power(self, user_id: str, server_id: str, action: str):
        try:
            user_id_int = int(user_id)
        except ValueError:
            raise HTTPException(status_code=400, detail="Invalid user ID format.")
            
        server_record = self.server_repo.get(server_id)
        
        if not server_record:
            raise HTTPException(status_code=404, detail="Server not registered in database.")
            
        if str(server_record.owner_id) != str(user_id):
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="You do not own this server.")

        if action == "stop":
            if not server_record.active_pod_name or not self.manager.get_container(server_record.active_pod_name):
                raise HTTPException(status_code=404, detail="Server instance not found or already stopped.")
                
            self.manager.stop_server(server_record.active_pod_name)
            self.server_repo.update(server_record, {"active_pod_name": None})

        elif action == "start":
            # 1. Async Credit Check via injected client
            credits = await self.identity_client.get_user_credits(user_id)
            if credits <= 1.0:
                raise HTTPException(status_code=402, detail="Insufficient credits.")
            
            if server_record.active_pod_name and self.manager.get_container(server_record.active_pod_name):
                return {"status": "already_running"}
            
            ephemeral_sidecar_token = secrets.token_hex(16)
            self.redis.set(f"sidecar_auth:{server_id}", ephemeral_sidecar_token)
            
            new_container = self.manager.create_server(
                game_id=server_record.game_id,
                user_id=user_id,
                logical_server_id=server_id,
                sidecar_token=ephemeral_sidecar_token,
                config_data=server_record.config
            )
            
            self.server_repo.update(server_record, {"active_pod_name": new_container.name})
        
        return {"result": "success", "status": "processing"}

    async def deploy_server(self, user_id: str, game_id: str, config: dict):
        if game_id == "valheim":
            try:
                config = ValheimConfigValidator(**config).dict()
            except Exception as e:
                raise HTTPException(status_code=400, detail=f"Invalid configuration: {str(e)}")
        
        try:
            user_id_int = int(user_id)
        except ValueError:
            raise HTTPException(status_code=400, detail="Invalid user ID format.")
        
        # 1. Async Credit Check via injected client
        credits = await self.identity_client.get_user_credits(user_id)
        if credits < 5.0:
            raise HTTPException(status_code=402, detail="Insufficient credits.")
        
        logical_server_id = str(uuid.uuid4())
        ephemeral_sidecar_token = secrets.token_hex(16)
        
        self.redis.set(f"sidecar_auth:{logical_server_id}", ephemeral_sidecar_token)
        
        new_container = self.manager.create_server(
            game_id=game_id,
            user_id=user_id,
            logical_server_id=logical_server_id,
            sidecar_token=ephemeral_sidecar_token,
            config_data=config
        )
        
        try:
            self.server_repo.create({
                "id": logical_server_id,
                "owner_id": user_id_int,
                "game_id": game_id,
                "config": config,
                "active_pod_name": new_container.name,
                "hourly_cost": 0.10
            })
        except Exception as e:
            raise HTTPException(status_code=500, detail=f"Failed to register server: {str(e)}")

        return {"status": "success", "server_id": logical_server_id}

File: services/fleet-manager/internal/service/server_service.go
package service

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"strconv"
	"strings"

	"fleet-manager/internal/client"
	"fleet-manager/internal/k8s"
	"fleet-manager/internal/model"
	"fleet-manager/internal/repository"

	"github.com/google/uuid"
	"github.com/redis/go-redis/v9"
)

type ServerService struct {
	repo           *repository.ServerRepository
	manager        *k8s.ServerManager
	redis          *redis.Client
	identityClient *client.IdentityClient
}

func NewServerService(
	repo *repository.ServerRepository,
	manager *k8s.ServerManager,
	redis *redis.Client,
	identityClient *client.IdentityClient,
) *ServerService {
	return &ServerService{
		repo:           repo,
		manager:        manager,
		redis:          redis,
		identityClient: identityClient,
	}
}

func (s *ServerService) ListServers(ctx context.Context, userID string) ([]model.ServerDetails, error) {
	servers, err := s.repo.GetByOwner(ctx, userID)
	if err != nil {
		return nil, err
	}

	var results []model.ServerDetails
	for _, srv := range servers {
		isOnline := false
		if srv.ActivePodName != nil && *srv.ActivePodName != "" {
			container := s.manager.GetContainer(*srv.ActivePodName)
			if container != nil && container.Status == "running" {
				isOnline = true
			}
		}

		displayName := getDisplayName(srv)

		results = append(results, model.ServerDetails{
			ID:      srv.ID,
			Name:    displayName,
			Status:  map[bool]string{true: "online", false: "offline"}[isOnline],
			CPU:     0,
			RAM:     0,
			Players: 0,
		})
	}

	return results, nil
}

func (s *ServerService) GetServerDetails(ctx context.Context, serverID string) (*model.ServerDetails, error) {
	srv, err := s.repo.Get(ctx, serverID)
	if err != nil {
		return nil, errors.New("server not found in database")
	}

	isOnline := false
	if srv.ActivePodName != nil && *srv.ActivePodName != "" {
		container := s.manager.GetContainer(*srv.ActivePodName)
		isOnline = container != nil && container.Status == "running"
	}

	var cpu, ram float64
	var players int

	stats, _ := s.redis.HGetAll(ctx, "server_stats:"+serverID).Result()
	if isOnline {
		cpu, _ = strconv.ParseFloat(stats["cpu"], 64)
		ram, _ = strconv.ParseFloat(stats["ram"], 64)
		players, _ = strconv.Atoi(stats["players"])
	}

	return &model.ServerDetails{
		ID:      srv.ID,
		Name:    getDisplayName(srv),
		Status:  map[bool]string{true: "online", false: "offline"}[isOnline],
		CPU:     cpu,
		RAM:     ram,
		Players: players,
	}, nil
}

func (s *ServerService) TogglePower(ctx context.Context, userID, serverID, action string) (map[string]string, error) {
	srv, err := s.repo.Get(ctx, serverID)
	if err != nil {
		return nil, errors.New("server not registered in database")
	}

	if srv.OwnerID != userID {
		return nil, errors.New("you do not own this server")
	}

	switch action {
	case "stop":
		if srv.ActivePodName == nil || s.manager.GetContainer(*srv.ActivePodName) == nil {
			return nil, errors.New("server instance not found or already stopped")
		}
		s.manager.StopServer(*srv.ActivePodName)
		s.repo.UpdateActivePod(ctx, srv.ID, nil)

	case "start":
		credits, err := s.identityClient.GetUserCredits(ctx, userID)
		if err != nil || credits <= 1.0 {
			return nil, errors.New("insufficient credits")
		}

		if srv.ActivePodName != nil && s.manager.GetContainer(*srv.ActivePodName) != nil {
			return map[string]string{"status": "already_running"}, nil
		}

		token := generateSecureToken()
		s.redis.Set(ctx, "sidecar_auth:"+serverID, token, 0)

		newContainer, err := s.manager.CreateServer(srv.GameID, userID, serverID, token, srv.Config)
		if err != nil {
			return nil, err
		}

		s.repo.UpdateActivePod(ctx, srv.ID, &newContainer.Name)
	default:
		return nil, errors.New("invalid action")
	}

	return map[string]string{"result": "success", "status": "processing"}, nil
}

func (s *ServerService) DeployServer(ctx context.Context, userID, gameID string, config map[string]interface{}) (map[string]string, error) {
	// 1. Validate Configs via JSON unmarshaling into strict structs
	configBytes, _ := json.Marshal(config)

	if gameID == "valheim" {
		var valheimCfg model.ValheimConfig
		if err := json.Unmarshal(configBytes, &valheimCfg); err != nil {
			return nil, fmt.Errorf("invalid configuration: %v", err)
		}
		if err := valheimCfg.Validate(); err != nil {
			return nil, fmt.Errorf("invalid configuration: %v", err)
		}
	} else if gameID == "test-server" {
		var testCfg model.TestServerConfig
		if err := json.Unmarshal(configBytes, &testCfg); err != nil {
			return nil, fmt.Errorf("invalid configuration: %v", err)
		}
		testCfg.SetupDefaults()
		// Re-marshal back to map to ensure defaults are saved
		newBytes, _ := json.Marshal(testCfg)
		json.Unmarshal(newBytes, &config)
	}

	// 2. Check Credits
	credits, err := s.identityClient.GetUserCredits(ctx, userID)
	if err != nil || credits < 5.0 {
		return nil, errors.New("insufficient credits")
	}

	// 3. Generate IDs and Tokens
	logicalServerID := uuid.New().String()
	token := generateSecureToken()
	s.redis.Set(ctx, "sidecar_auth:"+logicalServerID, token, 0)

	// 4. Deploy to Kubernetes
	newContainer, err := s.manager.CreateServer(gameID, userID, logicalServerID, token, config)
	if err != nil {
		return nil, err
	}

	// 5. Save to Database
	newServer := &model.Server{
		ID:            logicalServerID,
		OwnerID:       userID,
		GameID:        gameID,
		Config:        config,
		ActivePodName: &newContainer.Name,
		HourlyCost:    0.10,
	}

	if err := s.repo.Create(ctx, newServer); err != nil {
		return nil, fmt.Errorf("failed to register server: %v", err)
	}

	return map[string]string{"status": "success", "server_id": logicalServerID}, nil
}

// --- Helpers ---

func getDisplayName(srv *model.Server) string {
	if name, ok := srv.Config["VALHEIM_SERVER_NAME"].(string); ok && name != "" {
		return name
	}
	if name, ok := srv.Config["TEST_SERVER_NAME"].(string); ok && name != "" {
		return name
	}
	return strings.Title(srv.GameID) + " Server"
}

func generateSecureToken() string {
	bytes := make([]byte, 16)
	rand.Read(bytes)
	return hex.EncodeToString(bytes)
}